server/forge/sourcehut/sourcehut.go

```diff
@@ -0,0 +1,510 @@
+	if err != nil {
+		return err
+	}
+
+	// Note: we would only ever have registered one webhook
+	webhook := webhooks.Results[0]
+	_, err = git.UnregisterPushWebhook(client, ctx, webhook.Id)
+	return err
+}
+
+func (c *SourceHut) Branches(ctx context.Context, u *model.User, r *model.Repo, p *model.ListOptions) ([]string, error) {
+	// TODO: Pagination
+	if p.Page != 1 {
+		return nil, nil
+	}
+
+	client := c.newClientToken(ctx, c.gitURL, u.AccessToken)
+	owner, name, _ := strings.Cut(string(r.ForgeRemoteID), "/")
+
+	var branches []string
+	user, err := git.GetReferences(client, ctx, owner[1:], name, nil)
+	if err != nil {
+		return nil, nil
+	}
+
+	for _, ref := range user.Repository.References.Results {
+		if strings.HasPrefix(ref.Name, "refs/heads/") {
+			branches = append(branches, ref.Name[len("refs/heads/"):])
+		}
+	}
+
+	return branches, nil
+}
+
+func (c *SourceHut) BranchHead(ctx context.Context, u *model.User, r *model.Repo, branch string) (*model.Commit, error) {
+	client := c.newClientToken(ctx, c.gitURL, u.AccessToken)
+	owner, name, _ := strings.Cut(string(r.ForgeRemoteID), "/")
+	user, err := git.GetHead(client, ctx, owner[1:], name, "refs/heads/"+branch)
+	if err != nil {
+		return nil, err
+	}
+
+	target := user.Repository.Reference.Follow
+	return &model.Commit{
+		SHA:      target.Id,
+		ForgeURL: fmt.Sprintf("%s/%s/%s/commit/%s", c.gitURL, owner, name, target.Id),
+	}, nil
+}
+
+func (c *SourceHut) PullRequests(ctx context.Context, u *model.User, r *model.Repo, p *model.ListOptions) ([]*model.PullRequest, error) {
+	return nil, nil // TODO
+}
+
+func (c *SourceHut) Hook(ctx context.Context, r *http.Request) (*model.Repo, *model.Pipeline, error) {
+	type hookData struct {
+		Data struct {
+			Webhook *git.WebhookPayload `json:"webhook"`
+			Version *git.Version        `json:"version"`
+		} `json:"data"`
+	}
+
+	bytes, err := io.ReadAll(r.Body)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var hook hookData
+	err = json.Unmarshal(bytes, &hook)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var gitEvent *git.GitEvent
+	switch payload := hook.Data.Webhook.Value.(type) {
+	case *git.GitEvent:
+		gitEvent = payload
+	default:
+		log.Warn().Msg("Ignoring unknown webhook event")
+		return nil, nil, nil
+	}
+
+	repo := c.toRepo(gitEvent.Repository, hook.Data.Version)
+	pipeline := c.toPushPipeline(gitEvent)
+	return repo, pipeline, nil
+}
+
+func (c *SourceHut) OrgMembership(ctx context.Context, u *model.User, org string) (*model.OrgPerm, error) {
+	return &model.OrgPerm{
+		Member: true,
+		Admin:  true,
+	}, nil
+}
+
+func (c *SourceHut) Org(ctx context.Context, u *model.User, org string) (*model.Org, error) {
+	client := c.newClientToken(ctx, c.metaURL, u.AccessToken)
+	account, err := meta.FetchUser(client, ctx, org[1:])
+	if err != nil {
+		return nil, err
+	}
+	return &model.Org{
+		Name:    account.CanonicalName,
+		IsUser:  true,
+		Private: false,
+	}, nil
+}
```

Comment:
SourceHut does not (currently) support orgs. I did my best here to shim in the necessary code to account for that, but some extra attention here would be great.